lib/wallet/src/Cardano/Wallet/Transaction.hs

```diff
@@ -497,6 +502,28 @@ emptyWitnessCount = WitnessCount
     , bootstrap = 0
     }
 
+-- WitnessCount context is needed to differentiate verification keys present
+-- in native scripts.
+-- In shelley wallets they could be present due to only policy verification key.
+-- In multisig wallet they could stem from payment, policy and delegation roles,
+-- and as minting/burning and delegation support comes will be extended in additional
+-- data attached in SharedWalletCtx to differentiate that.
+-- WitnessCount is needed only during or after signing, in other phases it is not used.
+data WitnessCountCtx =
+    ShelleyWalletCtx KeyHash | SharedWalletCtx | AnyWitnessCountCtx
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+toKeyRole :: WitnessCountCtx -> ByteString -> KeyRole
+toKeyRole witCtx key = case witCtx of
+    ShelleyWalletCtx (KeyHash _ mypolicykey) ->
+        if key == mypolicykey then
+            Policy
+        else
+            Unknown
+    SharedWalletCtx -> Payment
+    AnyWitnessCountCtx -> error "toKeyRole was used in wrong WitnessCountCtx"
```

Comment:
yes, I was also thinking about this. what restrained me from doing this is that this WitnessCount is ONLY evaluated only when decode(Shared)Transaction is called. But we can now remove this partiality and ascribe Unknown! Will do it in coming commit

Reply:
> But we can now remove this partiality and ascribe Unknown!

Ok, fair enough.

Though to be fair I'm wondering if not adding the `WitnessCountCtx` argument is causing `fromCardanoTx` to do too much at once, and it would be neater to split anyway